# 01. Data Cleaning


In [37]:
import pandas as pd   # tables (DataFrames)
import ast            # turns text that looks like a list or dict into a real Python object

## 1. Load the raw datasets


In [38]:
# Load the raw files

customers = pd.read_csv('../data/raw/customers.csv')
offers = pd.read_csv('../data/raw/offers.csv')
events = pd.read_csv('../data/raw/events.csv')

In [39]:
# (rows, columns) of each table
print(customers.shape, offers.shape, events.shape)

(17000, 5) (10, 6) (306534, 4)


## 2. Clean the customers table

**What we check first:** column types, missing values and outliers (values that look wrong).




In [40]:
# first look at the customers table
customers.head()

,customer_id,became_member_on,gender,age,income
0,68be06ca386d4c31939f3a4f0e3dd783,20170212,NaN,118,NaN
1,0610b486422d4921ae7d2bf64640c50b,20170715,F,55,112000.0
2,38fe809add3b4fcf9315a9694bb96ff5,20180712,NaN,118,NaN
3,78afa995795e4d85b5d9ceeca43f5fef,20170509,F,75,100000.0
4,a03223e636434f42ac4c3df47e8bac43,20170804,NaN,118,NaN


In [41]:
customers.info()

# become_member_date is an object type, we will convert it to datetime type.


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 17000 entries, 0 to 16999
Data columns (total 5 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customer_id       17000 non-null  object 
 1   became_member_on  17000 non-null  int64  
 2   gender            14825 non-null  object 
 3   age               17000 non-null  int64  
 4   income            14825 non-null  float64
dtypes: float64(1), int64(2), object(2)
memory usage: 664.2+ KB


In [42]:
customers.describe()

#  the maximum age is 118, which could be an outlier.
# No zero/negative values in income and the max income is well withing a plausible range.

,became_member_on,age,income
count,1.700000e+04,17000.000000,14825.000000
mean,2.016703e+07,62.531412,65404.991568
std,1.167750e+04,26.738580,21598.299410
min,2.013073e+07,18.000000,30000.000000
25%,2.016053e+07,45.000000,49000.000000
50%,2.017080e+07,58.000000,64000.000000
75%,2.017123e+07,73.000000,80000.000000
max,2.018073e+07,118.000000,120000.000000


In [43]:

customers.isnull().sum()

# gender and income have missing values, we will explore them further.


customer_id            0
became_member_on       0
gender              2175
age                    0
income              2175
dtype: int64

### 2.1 Handle the incomplete profiles

**Finding:** every customer with `age = 118` also has missing gender and income. They are the same 2,175 rows (12.8% of customers). 118 is a placeholder, not a real age, so these are customers who never completed their profile.

**Decision:** drop them. This project compares customers by age and income, and we cannot do that for people with no real age or income.

**Alternative not chosen:** filling in a typical age and income. That would invent data for 12.8% of customers and distort every demographic comparison.

**Trade-off to remember:** our results describe customers with complete profiles only. Customers without a profile may behave differently.

In [44]:
# First check the frequency of age values.
print(customers['age'].value_counts().head())


# do the missing gender rows line up with age=118?
print((customers['age'] == 118).sum())
print(((customers['age'] == 118) & (customers['gender'].isna())).sum())


# both missing values and age = 118 are same rows, so they are supposed to the customers who never completed their profile.
# Since our whole project is about segmenting people by age and income,
# we can't use customers who don't have real age/income data — so we drop them.

age
118    2175
58      408
53      372
51      363
59      359
Name: count, dtype: int64
2175
2175


In [45]:
# drop the incomplete-profile customers (age = 118)
# .copy() makes an independent table, so later changes do not touch the raw one
clean_customers = customers[customers['age'] != 118].copy()

In [46]:
# Log exactly how many customers (and what %) were dropped, and why.

n_before, n_after = customers.shape[0], clean_customers.shape[0]
n_dropped = n_before - n_after
print(f"Customers before cleaning: {n_before}")
print(f"Customers after cleaning:  {n_after}")
print(f"Dropped (age=118 / incomplete profile): {n_dropped} ({n_dropped / n_before:.1%})")

Customers before cleaning: 17000
Customers after cleaning:  14825
Dropped (age=118 / incomplete profile): 2175 (12.8%)


In [47]:

# became_member_on is stored as a plain number like 20170715.
# We convert it into a real date so we can later calculate things
# like "how long has this person been a member" (tenure).
clean_customers['became_member_on'] = pd.to_datetime(clean_customers['became_member_on'], format='%Y%m%d')

print(clean_customers.shape)

(14825, 5)


In [48]:
# Check for fully duplicated customer rows (expected: 0)
n_dupes = clean_customers.duplicated().sum()
n_dupes

np.int64(0)

## 3. Clean the offers table
The offers table is small (10 offers) and has no missing values. The only issue is that `channels` is stored as text that *looks* like a list, e.g. `"['email', 'mobile']"`. We convert it to a real Python list so we can later check things like "was social media used for this offer?".

In [49]:
offers.head()

,offer_id,offer_type,difficulty,reward,duration,channels
0,ae264e3637204a6fb9bb56bc8210ddfd,bogo,10,10,7,"['email', 'mobile', 'social']"
1,4d5c57ea9a6940dd891ad53e9dbe8da0,bogo,10,10,5,"['web', 'email', 'mobile', 'social']"
2,3f207df678b143eea3cee63160fa8bed,informational,0,0,4,"['web', 'email', 'mobile']"
3,9b98b8c7a33c4b65b9aebfe6a799e6d9,bogo,5,5,7,"['web', 'email', 'mobile']"
4,0b1e1539f2cc45b7b9fa7c272da2e1d7,discount,20,5,10,"['web', 'email']"


In [50]:

offers.info()
offers.describe()

# The 'channels' column looks like a list (e.g. ['email','mobile'])
# but it's actually just TEXT — a string that looks like a list.
# ast.literal_eval reads that text and turns it into a real Python list,
# so we can later check things like "was social included in this offer"

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10 entries, 0 to 9
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   offer_id    10 non-null     object
 1   offer_type  10 non-null     object
 2   difficulty  10 non-null     int64 
 3   reward      10 non-null     int64 
 4   duration    10 non-null     int64 
 5   channels    10 non-null     object
dtypes: int64(3), object(3)
memory usage: 612.0+ bytes


,difficulty,reward,duration
count,10.000000,10.000000,10.000000
mean,7.700000,4.200000,6.500000
std,5.831905,3.583915,2.321398
min,0.000000,0.000000,3.000000
25%,5.000000,2.000000,5.000000
50%,8.500000,4.000000,7.000000
75%,10.000000,5.000000,7.000000
max,20.000000,10.000000,10.000000


In [51]:
# ast.literal_eval safely reads text like "['email', 'mobile']" and returns a real list
def get_channels_list(text):
    return ast.literal_eval(text)

offers['channels_list'] = offers['channels'].apply(get_channels_list)

# check: the first offer's channels, and confirm the type is now a list
print(offers['channels_list'][0])
print(type(offers['channels_list'][0]))

['email', 'mobile', 'social']
<class 'list'>


## 4. Clean the events table

The events table is a log of everything customers did during the 30 days. Each row is one action: `offer received`, `offer viewed`, `offer completed` or `transaction`.

The `value` column holds different information depending on the event type, stored as text that looks like a dictionary:

| Event | What `value` contains |
|---|---|
| offer received / offer viewed | offer id |
| offer completed | offer id and reward |
| transaction | dollar amount |

The next few cells explore this structure before we clean it.

In [52]:
events.head()
events.tail()

# The 'value' column is also just TEXT that looks like a dictionary.
# Depending on the event type, it holds different information:
#   - offer received / offer viewed  -> holds an offer id
#   - offer completed                -> holds an offer id AND a reward
#   - transaction                    -> holds a dollar amount
# We need to pull each of these out into its own real column.

,customer_id,event,value,time
306529,b3a1272bc9904337b331bf348c3e8c17,transaction,{'amount': 1.5899999999999999},714
306530,68213b08d99a4ae1b0dcb72aebd9aa35,transaction,{'amount': 9.53},714
306531,a00058cf10334a308c68e7631c529907,transaction,{'amount': 3.61},714
306532,76ddbd6576844afe811f1a3c0fbb5bec,transaction,{'amount': 3.5300000000000002},714
306533,c02b10e8752c4d8e9b73f918558531f7,transaction,{'amount': 4.05},714


In [53]:
events.info()
events.describe()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 306534 entries, 0 to 306533
Data columns (total 4 columns):
 #   Column       Non-Null Count   Dtype 
---  ------       --------------   ----- 
 0   customer_id  306534 non-null  object
 1   event        306534 non-null  object
 2   value        306534 non-null  object
 3   time         306534 non-null  int64 
dtypes: int64(1), object(3)
memory usage: 9.4+ MB


,time
count,306534.000000
mean,366.382940
std,200.326314
min,0.000000
25%,186.000000
50%,408.000000
75%,528.000000
max,714.000000


In [54]:
events['value'].head() 
events['value'].tail() 

306529    {'amount': 1.5899999999999999}
306530                  {'amount': 9.53}
306531                  {'amount': 3.61}
306532    {'amount': 3.5300000000000002}
306533                  {'amount': 4.05}
Name: value, dtype: object

In [55]:
events[events['event'] == 'transaction']['value'].head()

12654    {'amount': 0.8300000000000001}
12657                 {'amount': 34.56}
12659                 {'amount': 13.23}
12670                 {'amount': 19.51}
12671                 {'amount': 18.97}
Name: value, dtype: object

In [56]:
events[events['event'] == 'offer completed'][['event', 'value']].head(10)

,event,value
12658,offer completed,{'offer_id': '2906b810c7d4411798c6938adc9daaa5...
12672,offer completed,{'offer_id': 'fafdcd668e3743c1bb461111dcafc2a4...
12679,offer completed,{'offer_id': '9b98b8c7a33c4b65b9aebfe6a799e6d9...
12692,offer completed,{'offer_id': 'ae264e3637204a6fb9bb56bc8210ddfd...
12697,offer completed,{'offer_id': '4d5c57ea9a6940dd891ad53e9dbe8da0...
12717,offer completed,{'offer_id': '4d5c57ea9a6940dd891ad53e9dbe8da0...
12721,offer completed,{'offer_id': '2298d6c36e964ae4a3e7e9706d1fb8c2...
12744,offer completed,{'offer_id': 'f19421c1d4aa40978ebb69ca19b0e20d...
12764,offer completed,{'offer_id': '2906b810c7d4411798c6938adc9daaa5...
12767,offer completed,{'offer_id': '4d5c57ea9a6940dd891ad53e9dbe8da0...


In [57]:
events['event'].value_counts()

event
transaction        138953
offer received      76277
offer viewed        57725
offer completed     33579
Name: count, dtype: int64

### 4.a Extract the fields into their own columns
We turn the text in `value` into a real dictionary and pull out `offer_id`, `amount` and `reward_value` as separate columns. A row that does not have a field gets an empty value (for example, transactions have no offer id). The original `value` column is dropped afterwards.

The key name can differ between event types (`'offer id'` vs `'offer_id'`), so the function checks both.

In [58]:


# This function grabs the offer id out of the text, if it has one.
# .get() is used instead of ['offer id'] because not every row has
# an offer id (transactions don't) — .get() just returns empty (None)
# instead of crashing when the key isn't there.


def get_offer_id(value_text):
    d = ast.literal_eval(value_text)          # turn text into a real dict
    return d.get('offer id') or d.get('offer_id')

# This function grabs the dollar amount, only present on transactions.
def get_amount(value_text):
    d = ast.literal_eval(value_text)
    return d.get('amount')

# This function grabs the reward dollar value, only present on completions.
def get_reward(value_text):
    d = ast.literal_eval(value_text)
    return d.get('reward')

In [59]:
# Run each function on every row of the 'value' column,
# creating three brand new, clean columns.
events['offer_id'] = events['value'].apply(get_offer_id)
events['amount'] = events['value'].apply(get_amount)
events['reward_value'] = events['value'].apply(get_reward)


In [60]:
# spot-check: transaction rows should have an amount and no offer_id
print(events[events['event'] == 'transaction'].head())

                            customer_id        event  \
12654  02c083884c7d45b39cc68e1314fec56c  transaction   
12657  9fa9ae8f57894cc9a3b8a9bbe0fc1b2f  transaction   
12659  54890f68699049c2a04d415abc25e717  transaction   
12670  b2f1cd155b864803ad8334cdf13c4bd2  transaction   
12671  fe97aa22dd3e48c8b143116a8403dd52  transaction   

                                value  time offer_id  amount  reward_value  
12654  {'amount': 0.8300000000000001}     0     None    0.83           NaN  
12657               {'amount': 34.56}     0     None   34.56           NaN  
12659               {'amount': 13.23}     0     None   13.23           NaN  
12670               {'amount': 19.51}     0     None   19.51           NaN  
12671               {'amount': 18.97}     0     None   18.97           NaN  


In [61]:
# spot-check: completed rows should have an offer_id and a reward_value
print(events[events['event'] == 'offer completed'].head())

                            customer_id            event  \
12658  9fa9ae8f57894cc9a3b8a9bbe0fc1b2f  offer completed   
12672  fe97aa22dd3e48c8b143116a8403dd52  offer completed   
12679  629fc02d56414d91bca360decdfa9288  offer completed   
12692  676506bad68e4161b9bbaffeb039626b  offer completed   
12697  8f7dd3b2afe14c078eb4f6e6fe4ba97d  offer completed   

                                                   value  time  \
12658  {'offer_id': '2906b810c7d4411798c6938adc9daaa5...     0   
12672  {'offer_id': 'fafdcd668e3743c1bb461111dcafc2a4...     0   
12679  {'offer_id': '9b98b8c7a33c4b65b9aebfe6a799e6d9...     0   
12692  {'offer_id': 'ae264e3637204a6fb9bb56bc8210ddfd...     0   
12697  {'offer_id': '4d5c57ea9a6940dd891ad53e9dbe8da0...     0   

                               offer_id  amount  reward_value  
12658  2906b810c7d4411798c6938adc9daaa5     NaN           2.0  
12672  fafdcd668e3743c1bb461111dcafc2a4     NaN           2.0  
12679  9b98b8c7a33c4b65b9aebfe6a799e6d9     NaN   

In [62]:
# Keep only events that belong to customers who survived the customer cleaning above
# (this removes the events of the incomplete-profile customers)
events_clean = events[events['customer_id'].isin(clean_customers['customer_id'])].copy()

In [63]:
# We don't need the original messy 'value' column anymore — 
# we've already extracted everything useful out of it.
events_clean = events_clean.drop(columns=['value'])



In [64]:
# Log how many events were removed because their customer was dropped
events_before, events_after = events.shape[0], events_clean.shape[0]
events_dropped = events_before - events_after

print(f"Events before cleaning: {events_before}")
print(f"Events after cleaning:  {events_after}")

Events before cleaning: 306534
Events after cleaning:  272762


### 4.b Data quality checks

Before saving, we check that the extraction in 4.a worked on every row (not just the few rows we looked at with `.head()`), and that events point to real offers. Three checks follow:
1. each event type fills in exactly the fields it should,
2. every `offer_id` in events exists in the offers table,
3. transaction amounts are positive.

In [65]:
# Check 1: every event type should populate exactly the fields we expect
# from it, and nothing else. If this doesn't hold, get_offer_id/get_amount/
# get_reward extracted something wrong.
null_pattern = events_clean.groupby('event')[['offer_id', 'amount', 'reward_value']].apply(
    lambda g: g.isnull().mean()
)
print(null_pattern)

# Expected:
#   transaction        -> offer_id null 100%, amount null 0%,   reward_value null 100%
#   offer received      -> offer_id null 0%,   amount null 100%, reward_value null 100%
#   offer viewed        -> offer_id null 0%,   amount null 100%, reward_value null 100%
#   offer completed      -> offer_id null 0%,   amount null 100%, reward_value null 0%

                 offer_id  amount  reward_value
event                                          
offer completed       0.0     1.0           0.0
offer received        0.0     1.0           1.0
offer viewed          0.0     1.0           1.0
transaction           1.0     0.0           1.0


In [66]:
# Check 2: referential integrity — every non-null offer_id in events
# should exist in the offers table. Orphaned IDs would silently break
# any join we do later (offer type, difficulty, reward, duration).
event_offer_ids = events_clean['offer_id'].dropna().unique()
orphaned_ids = set(event_offer_ids) - set(offers['offer_id'])
print(f"Distinct offer_ids referenced in events: {len(event_offer_ids)}")
print(f"Orphaned offer_ids (not in offers table): {len(orphaned_ids)}")

Distinct offer_ids referenced in events: 10
Orphaned offer_ids (not in offers table): 0


In [67]:
# Check 3: transaction amounts should be strictly positive — a zero or
# negative "purchase" would be a data error and would distort revenue
# and average-spend metrics.
tx = events_clean[events_clean['event'] == 'transaction']
print("Transactions with null amount:", tx['amount'].isnull().sum())
print("Transactions with amount <= 0:", (tx['amount'] <= 0).sum())
print(tx['amount'].describe())

Transactions with null amount: 0
Transactions with amount <= 0: 0
count    123957.000000
mean         13.996325
std          31.749420
min           0.050000
25%           3.660000
50%          10.800000
75%          19.130000
max        1062.280000
Name: amount, dtype: float64


**Why the extraction matters:** the `value` column was first converted from a string into a dictionary. Then the relevant fields were extracted into separate columns (`offer_id`, `amount`, `reward_value`), so each type of information is clearly represented. This makes it possible to link events to offers, analyse customer spending, and follow the full customer journey.

### 4.c Remove duplicate events

The events table contains fully duplicated rows (374 in this data), found by comparing entire rows. We check for them after the consistency and integrity checks above, so we know the extraction is trustworthy first.

Each row is one customer action at one time, so identical rows are not valid. For example, two identical transaction rows for the same customer, amount and timestamp would double-count revenue. We remove them to prevent double counting and keep KPIs accurate.

In [68]:
# count exact duplicate rows, then keep one copy of each
n_before_dedup = events_clean.shape[0]
n_dupes = events_clean.duplicated().sum()
print(f"Fully duplicated rows found: {n_dupes}")

events_clean = events_clean.drop_duplicates().reset_index(drop=True)

# check: the difference should equal the number of duplicates found
n_after_dedup = events_clean.shape[0]

print(f"Events before dedup: {n_before_dedup}")
print(f"Events after dedup:  {n_after_dedup}")
print(f"Dropped: {n_before_dedup - n_after_dedup}")

Fully duplicated rows found: 374
Events before dedup: 272762
Events after dedup:  272388
Dropped: 374


### 4.d Add time in days
The `time` column counts **hours** since the start of the 30-day period (0 to 714). We add a `time_days` column because offer durations are given in days, which makes them easier to compare at a glance.

The original hours column is kept: later notebooks use hours for exact comparisons, because events are recorded in whole hours.

In [69]:
# The 'time' column counts hours since the start of the 30-day period
# (0 to 714 hours). We convert it to days.
events_clean['time_days'] = events_clean['time'] / 24
events_clean.tail()

,customer_id,event,time,offer_id,amount,reward_value,time_days
272383,24f56b5e1849462093931b164eb803b5,offer completed,714,fafdcd668e3743c1bb461111dcafc2a4,NaN,2.0,29.75
272384,b3a1272bc9904337b331bf348c3e8c17,transaction,714,None,1.59,NaN,29.75
272385,68213b08d99a4ae1b0dcb72aebd9aa35,transaction,714,None,9.53,NaN,29.75
272386,a00058cf10334a308c68e7631c529907,transaction,714,None,3.61,NaN,29.75
272387,76ddbd6576844afe811f1a3c0fbb5bec,transaction,714,None,3.53,NaN,29.75


### 4.e Note on saving to CSV

CSV can only store plain text, so two columns will lose their Python type on save and need to be re-parsed when reloaded in later notebooks:
- `became_member_on` (datetime) will come back as a string, so re-run `pd.to_datetime(...)`.
- `channels_list` (Python list) will come back as a string that *looks* like a list, e.g. `"['web', 'email']"`, so re-run `ast.literal_eval(...)` before using it.

Also note: `offers.reward` is the reward *amount promised* by the offer, while `events.reward_value` is the reward *actually paid out* on a specific completion event. Keep these separate in later analysis, because they answer different questions.

## 5. Save the cleaned datasets
The cleaned tables are saved to `data/cleaned/` and used by every later notebook. The raw files are left untouched.

In [70]:
# Save the cleaned tables (index=False so the row numbers are not written as an extra column)
clean_customers.to_csv('../data/cleaned/customers_clean.csv', index=False)
offers.to_csv('../data/cleaned/offers_clean.csv', index=False)
events_clean.to_csv('../data/cleaned/events_clean.csv', index=False)

# summary of what changed
print("Cleaning done.")
print("customers:", clean_customers.shape, f"({n_dropped} dropped, {n_dropped/n_before:.1%})")
print("offers:", offers.shape)
print("events:", events_clean.shape, f"({n_dupes} duplicate rows dropped)")

Cleaning done.
customers: (14825, 5) (2175 dropped, 12.8%)
offers: (10, 7)
events: (272388, 7) (374 duplicate rows dropped)


## 6. Load the cleaned data into PostgreSQL
The cleaned tables are also loaded into a PostgreSQL database, so the SQL analysis in `sql/` can run on exactly the same data.

Connection details (user, password, host, port, database) are read from a `.env` file, so no passwords are stored in the notebook. The `.env` file is listed in `.gitignore`.

**Before running:** the connection cell uses `os.getenv`, so `import os` and `load_dotenv()` (currently commented out in the cell below) need to be active.

In [71]:
#install packages
!pip install sqlalchemy psycopg2-binary pandas

# !pip install python-dotenv
#import os
#from dotenv import load_dotenv

#load_dotenv()  # loads variables from .env


Defaulting to user installation because normal site-packages is not writeable



[notice] A new release of pip is available: 26.0 -> 26.2.1
[notice] To update, run: python.exe -m pip install --upgrade pip


In [72]:
# connect jupyter to PostgreSQL
from sqlalchemy import create_engine

# Databse connection details
username = os.getenv("PG_USER")
password = os.getenv("PG_PASSWORD")
host = os.getenv("PG_HOST")
port = os.getenv("PG_PORT")
database = os.getenv("PG_DATABASE")

# Create PostgreSQL engine
engine = create_engine(f"postgresql+psycopg2://{username}:{password}@{host}:{port}/{database}")

# test connection
with engine.connect() as conn:
    print("Connected to PostgreSQL!")

NameError: name 'os' is not defined

In [ ]:
# load dataframe to Postgres
customers = pd.read_csv("../data/cleaned/customers_clean.csv")
offers = pd.read_csv("../data/cleaned/offers_clean.csv")
events = pd.read_csv("../data/cleaned/events_clean.csv")

# Load into public schema/table (auto-creates table)
customers.to_sql("customers", engine, schema="public", if_exists="replace", index=False)
offers.to_sql("offers", engine, schema="public", if_exists="replace", index=False)
events.to_sql("events", engine, schema="public", if_exists="replace", index=False)

print("Loaded into PostgreSQL: database _ clean datasets")